thrust/system/detail/generic/shuffle.inl

```diff
@@ -0,0 +1,208 @@
+  uint64_t left_side_bits;
+  uint64_t right_side_mask;
+  uint64_t left_side_mask;
+  uint64_t key[num_rounds];
+};
+
+struct key_flag_tuple {
+  uint64_t key;
+  uint64_t flag;
+};
+
+// scan only flags
+struct key_flag_scan_op {
+  __host__ __device__ key_flag_tuple operator()(const key_flag_tuple& a,
+                                                const key_flag_tuple& b) {
+    return {b.key, a.flag + b.flag};
+  }
+};
+
+struct construct_key_flag_op {
+  uint64_t m;
+  feistel_bijection bijection;
+  __host__ __device__ construct_key_flag_op(uint64_t m,
+                                            feistel_bijection bijection)
+      : m(m), bijection(bijection) {}
+  __host__ __device__ key_flag_tuple operator()(uint64_t idx) {
+    auto gather_key = bijection(idx);
+    return key_flag_tuple{gather_key, gather_key < m};
+  }
+};
+
+template <typename InputIterT, typename OutputIterT>
+struct write_output_op {
+  uint64_t m;
+  InputIterT in;
+  OutputIterT out;
+  // flag contains inclusive scan of valid keys
+  // perform gather using valid keys
+  __thrust_exec_check_disable__
+  __host__ __device__ size_t operator()(key_flag_tuple x) {
+    if (x.key < m) {
+      // -1 because inclusive scan
+      out[x.flag - 1] = in[x.key];
+    }
+    return 0;  // Discarded
+  }
+};
+
+template <typename ExecutionPolicy, typename RandomIterator, typename URBG>
+__host__ __device__ void shuffle(
+    thrust::execution_policy<ExecutionPolicy>& exec, RandomIterator first,
+    RandomIterator last, URBG&& g) {
+  typedef
+      typename thrust::iterator_traits<RandomIterator>::value_type InputType;
+
+  // copy input to temp buffer
+  thrust::detail::temporary_array<InputType, ExecutionPolicy> temp(exec, first,
+                                                                   last);
+  thrust::shuffle_copy(exec, temp.begin(), temp.end(), first, g);
+}
+
+template <typename ExecutionPolicy, typename RandomIterator,
+          typename OutputIterator, typename URBG>
+__host__ __device__ void shuffle_copy(
+    thrust::execution_policy<ExecutionPolicy>& exec, RandomIterator first,
+    RandomIterator last, OutputIterator result, URBG&& g) {
+  // m is the length of the input
+  // we have an available bijection of length n via a feistel cipher
+  size_t m = last - first;
+  feistel_bijection bijection(m, g);
+  uint64_t n = bijection.nearest_power_of_two();
+
+  // perform stream compaction over length n bijection to get length m
+  // pseudorandom bijection over the original input
+  thrust::counting_iterator<uint64_t> indices(0);
+  thrust::transform_iterator<construct_key_flag_op, decltype(indices),
+                             key_flag_tuple>
+      key_flag_it(indices, construct_key_flag_op(m, bijection));
+  write_output_op<RandomIterator, decltype(result)> write_functor{m, first,
+                                                                  result};
+  auto gather_output_it = thrust::make_transform_output_iterator(
```

Comment:
Since the implementation uses C++11 (`auto`, `decltype`, maybe others?), the `thrust/shuffle.h` header should have a C++11 guard or have the C++11 bits removed. See the beginning of `thrust/optional.h` for an example of how the C++11 guards work.